File: samples/dnn/text_detection.cpp
/*
    Text detection model: https://github.com/argman/EAST
    Download link: https://www.dropbox.com/s/r2ingd0l3zt8hxs/frozen_east_text_detection.tar.gz?dl=1

    Text recognition models can be downloaded directly here:
    Download link: https://drive.google.com/drive/folders/1cTbQ3nuZG-EKWak6emD_s8_hHXWz7lAr?usp=sharing
    and doc/tutorials/dnn/dnn_text_spotting/dnn_text_spotting.markdown

    How to convert from pb to onnx:
    Using classes from here: https://github.com/meijieru/crnn.pytorch/blob/master/models/crnn.py
    import torch
    from models.crnn import CRNN
    model = CRNN(32, 1, 37, 256)
    model.load_state_dict(torch.load('crnn.pth'))
    dummy_input = torch.randn(1, 1, 32, 100)
    torch.onnx.export(model, dummy_input, "crnn.onnx", verbose=True)

    For more information, please refer to doc/tutorials/dnn/dnn_text_spotting/dnn_text_spotting.markdown and doc/tutorials/dnn/dnn_OCR/dnn_OCR.markdown
*/
#include <iostream>
#include <fstream>

#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/dnn.hpp>

using namespace cv;
using namespace cv::dnn;

const char* keys =
    "{ help  h              | | Print help message. }"
    "{ input i              | | Path to input image or video file. Skip this argument to capture frames from a camera.}"
    "{ detModel dmp         | | Path to a binary .pb file contains trained detector network.}"
    "{ width                | 320 | Preprocess input image by resizing to a specific width. It should be a multiple of 32. }"
    "{ height               | 320 | Preprocess input image by resizing to a specific height. It should be a multiple of 32. }"
    "{ thr                  | 0.5 | Confidence threshold. }"
    "{ nms                  | 0.4 | Non-maximum suppression threshold. }"
    "{ recModel rmp         | | Path to a binary .onnx file contains trained CRNN text recognition model. "
        "Download links are provided in doc/tutorials/dnn/dnn_text_spotting/dnn_text_spotting.markdown}"
    "{ RGBInput rgb         |0| 0: imread with flags=IMREAD_GRAYSCALE; 1: imread with flags=IMREAD_COLOR. }"
    "{ vocabularyPath vp    | alphabet_36.txt | Path to benchmarks for evaluation. "
        "Download links are provided in doc/tutorials/dnn/dnn_text_spotting/dnn_text_spotting.markdown}";

void fourPointsTransform(const Mat& frame, const Point2f vertices[], Mat& result);

int main(int argc, char** argv)
{
    // Parse command line arguments.
    CommandLineParser parser(argc, argv, keys);
    parser.about("Use this script to run TensorFlow implementation (https://github.com/argman/EAST) of "
                 "EAST: An Efficient and Accurate Scene Text Detector (https://arxiv.org/abs/1704.03155v2)");
    if (argc == 1 || parser.has("help"))
    {
        parser.printMessage();
        return 0;
    }

    float confThreshold = parser.get<float>("thr");
    float nmsThreshold = parser.get<float>("nms");
    int width = parser.get<int>("width");
    int height = parser.get<int>("height");
    int imreadRGB = parser.get<int>("RGBInput");
    String detModelPath = parser.get<String>("detModel");
    String recModelPath = parser.get<String>("recModel");
    String vocPath = parser.get<String>("vocabularyPath");

    if (!parser.check())
    {
        parser.printErrors();
        return 1;
    }

    // Load networks.
    CV_Assert(!detModelPath.empty() && !recModelPath.empty());
    TextDetectionModel_EAST detector(detModelPath);
    detector.setConfidenceThreshold(confThreshold)
            .setNMSThreshold(nmsThreshold);

    TextRecognitionModel recognizer(recModelPath);

    // Load vocabulary
    CV_Assert(!vocPath.empty());
    std::ifstream vocFile;
    vocFile.open(samples::findFile(vocPath));
    CV_Assert(vocFile.is_open());
    String vocLine;
    std::vector<String> vocabulary;
    while (std::getline(vocFile, vocLine)) {
        vocabulary.push_back(vocLine);
    }
    recognizer.setVocabulary(vocabulary);
    recognizer.setDecodeType("CTC-greedy");

    // Parameters for Recognition
    double recScale = 1.0 / 127.5;
    Scalar recMean = Scalar(127.5, 127.5, 127.5);
    Size recInputSize = Size(100, 32);
    recognizer.setInputParams(recScale, recInputSize, recMean);

    // Parameters for Detection
    double detScale = 1.0;
    Size detInputSize = Size(width, height);
    Scalar detMean = Scalar(123.68, 116.78, 103.94);
    bool swapRB = true;
    detector.setInputParams(detScale, detInputSize, detMean, swapRB);

    // Open a video file or an image file or a camera stream.
    VideoCapture cap;
    bool openSuccess = parser.has("input") ? cap.open(parser.get<String>("input")) : cap.open(0);
    CV_Assert(openSuccess);

    static const std::string kWinName = "EAST: An Efficient and Accurate Scene Text Detector";

    Mat frame;
    while (waitKey(1) < 0)
    {
        cap >> frame;
        if (frame.empty())
        {
            waitKey();
            break;
        }

        std::cout << frame.size << std::endl;

        // Detection
        std::vector< std::vector<Point> > detResults;
        detector.detect(frame, detResults);
        Mat frame2 = frame.clone();
        if (detResults.size() > 0) {
            // Text Recognition
            Mat recInput;
            if (!imreadRGB) {
                cvtColor(frame, recInput, cv::COLOR_BGR2GRAY);
            } else {
                recInput = frame;
            }
            std::vector< std::vector<Point> > contours;
            for (uint i = 0; i < detResults.size(); i++)
            {
                const auto& quadrangle = detResults[i];
                CV_CheckEQ(quadrangle.size(), (size_t)4, "");

                contours.emplace_back(quadrangle);

                std::vector<Point2f> quadrangle_2f;
                for (int j = 0; j < 4; j++)
                    quadrangle_2f.emplace_back(quadrangle[j]);

                Mat cropped;
                fourPointsTransform(recInput, &quadrangle_2f[0], cropped);

                std::string recognitionResult = recognizer.recognize(cropped);
                std::cout << i << ": '" << recognitionResult << "'" << std::endl;

                putText(frame2, recognitionResult, quadrangle[3], FONT_HERSHEY_SIMPLEX, 1.5, Scalar(0, 0, 255), 2);
            }
            polylines(frame2, contours, true, Scalar(0, 255, 0), 2);
        }
        imshow(kWinName, frame2);
    }
    return 0;
}

void fourPointsTransform(const Mat& frame, const Point2f vertices[], Mat& result)
{
    const Size outputSize = Size(100, 32);

    Point2f targetVertices[4] = {
        Point(0, outputSize.height - 1),
        Point(0, 0), Point(outputSize.width - 1, 0),
        Point(outputSize.width - 1, outputSize.height - 1)
    };
    Mat rotationMatrix = getPerspectiveTransform(vertices, targetVertices);

    warpPerspective(frame, result, rotationMatrix, outputSize);
}

File: samples/dnn/text_detection.py
'''
    Text detection model (EAST): https://github.com/argman/EAST
    Download link for EAST model: https://www.dropbox.com/s/r2ingd0l3zt8hxs/frozen_east_text_detection.tar.gz?dl=1

    DB detector model:
    https://drive.google.com/uc?export=download&id=17_ABp79PlFt9yPCxSaarVc_DKTmrSGGf

    CRNN Text recognition model sourced from: https://github.com/meijieru/crnn.pytorch
    How to convert from .pb to .onnx:
    Using classes from: https://github.com/meijieru/crnn.pytorch/blob/master/models/crnn.py

    Additional converted ONNX text recognition models available for direct download:
    Download link: https://drive.google.com/drive/folders/1cTbQ3nuZG-EKWak6emD_s8_hHXWz7lAr?usp=sharing
    These models are taken from: https://github.com/clovaai/deep-text-recognition-benchmark

    Importing and using the CRNN model in PyTorch:
    import torch
    from models.crnn import CRNN

    model = CRNN(32, 1, 37, 256)
    model.load_state_dict(torch.load('crnn.pth'))
    dummy_input = torch.randn(1, 1, 32, 100)
    torch.onnx.export(model, dummy_input, "crnn.onnx", verbose=True)

    Usage: python text_detection.py DB --ocr=<path to recognition model>

'''
import os
import cv2
import argparse
import numpy as np
from common import *

def help():
    print(
        '''
        Use this script for Text Detection and Recognition using OpenCV.

        Firstly, download required models using `download_models.py` (if not already done). Set environment variable OPENCV_DOWNLOAD_CACHE_DIR to specify where models should be downloaded. Also, point OPENCV_SAMPLES_DATA_PATH to opencv/samples/data.
        To run:
        Example: ./example_dnn_text_detection modelName(i.e. DB or East) --ocr=<path to ResNet_CTC.onnx>\

        Model path can also be specified using --model argument.
        '''
    )

############ Add argument parser for command line arguments ############
def get_args_parser():
    parser = argparse.ArgumentParser(add_help=False)
    parser.add_argument('--input', default='right.jpg',
                        help='Path to input image or video file. Skip this argument to capture frames from a camera.')
    parser.add_argument('--zoo', default=os.path.join(os.path.dirname(os.path.abspath(__file__)), 'models.yml'),
                        help='An optional path to file with preprocessing parameters.')
    parser.add_argument('--ocr',
                        help="Path to a .onnx file contains trained recognition network", required=True)
    parser.add_argument('--thr', type=float, default=0.5,
                        help='Confidence threshold.')
    parser.add_argument('--nms', type=float, default=0.4,
                        help='Non-maximum suppression threshold.')
    parser.add_argument('--binary_threshold', type=float, default=0.3,
                        help='Confidence threshold for the binary map in DB detector. ')
    parser.add_argument('--polygon_threshold', type=float, default=0.5,
                        help='Confidence threshold for polygons in DB detector.')
    parser.add_argument('--max_candidate', type=int, default=200,
                        help='Max candidates for polygons in DB detector.')
    parser.add_argument('--unclip_ratio', type=float, default=2.0,
                        help='Unclip ratio for DB detector.')
    parser.add_argument('--vocabulary_path', default='alphabet_36.txt',
                        help='Path to vocabulary file.')
    args, _ = parser.parse_known_args()
    add_preproc_args(args.zoo, parser, 'text_detection')
    parser = argparse.ArgumentParser(parents=[parser],
                                        description='Text Detection and Recognition using OpenCV.',
                                        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    return parser.parse_args()

def fourPointsTransform(frame, vertices):
    vertices = np.asarray(vertices)
    outputSize = (100, 32)
    targetVertices = np.array([
        [0, outputSize[1] - 1],
        [0, 0],
        [outputSize[0] - 1, 0],
        [outputSize[0] - 1, outputSize[1] - 1]], dtype="float32")

    rotationMatrix = cv2.getPerspectiveTransform(vertices, targetVertices)
    result = cv2.warpPerspective(frame, rotationMatrix, outputSize)
    return result

def main():
    args = get_args_parser()
    print(args.alias)

    if args.alias is None or hasattr(args, 'help'):
        help()
        exit(1)

    args.model = findModel(args.model, args.sha1)
    args.input = findFile(args.input)
    args.vocabulary_path = findFile(args.vocabulary_path)

    frame = cv2.imread(args.input)

    if(args.alias == "DB"):
        # DB Detector initialization
        detector = cv2.dnn_TextDetectionModel_DB(args.model)
        detector.setBinaryThreshold(args.binary_threshold)
        detector.setPolygonThreshold(args.polygon_threshold)
        detector.setUnclipRatio(args.unclip_ratio)
        detector.setMaxCandidates(args.max_candidate)
        # Setting input parameters specific to the DB model
        detector.setInputParams(scale=args.scale, size=(args.width, args.height), mean=args.mean)
        # Performing text detection
        detResults = detector.detect(frame)
    elif(args.alias == "East"):
        # EAST Detector initialization
        detector = cv2.dnn_TextDetectionModel_EAST(args.model)
        detector.setConfidenceThreshold(args.thr)
        detector.setNMSThreshold(args.nms)
        # Setting input parameters specific to EAST model
        detector.setInputParams(scale=args.scale, size=(args.width, args.height), mean=args.mean, swapRB=True)
        # Perfroming text detection
        detResults = detector.detect(frame)

    # Open the vocabulary file and read lines into a list
    with open(args.vocabulary_path, 'r') as voc_file:
        vocabulary = [line.strip() for line in voc_file]

    # Initialize the text recognition model with the specified model path
    recognizer = cv2.dnn_TextRecognitionModel(args.ocr)

    # Set the vocabulary for the model
    recognizer.setVocabulary(vocabulary)

    # Set the decoding method to 'CTC-greedy'
    recognizer.setDecodeType("CTC-greedy")

    recScale = 1.0 / 127.5
    recMean = (127.5, 127.5, 127.5)
    recInputSize = (100, 32)
    recognizer.setInputParams(scale=recScale, size=recInputSize, mean=recMean)

    if len(detResults) > 0:
        recInput = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY) if not args.rgb else frame.copy()
        contours = []

        for i, (quadrangle, _) in enumerate(zip(detResults[0], detResults[1])):
            if isinstance(quadrangle, np.ndarray):
                quadrangle = np.array(quadrangle).astype(np.float32)

                if quadrangle is None or len(quadrangle) != 4:
                    print("Skipping a quadrangle with incorrect points or transformation failed.")
                    continue

                contours.append(np.array(quadrangle, dtype=np.int32))
                cropped = fourPointsTransform(recInput, quadrangle)
                recognitionResult = recognizer.recognize(cropped)
                print(f"{i}: '{recognitionResult}'")

                try:
                    text_origin = (int(quadrangle[3][0]), int(quadrangle[3][1]))
                    cv2.putText(frame, recognitionResult, text_origin, cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 0, 255), 2)
                except Exception as e:
                    print("Failed to write text on the frame:", e)
            else:
                print("Skipping a detection with invalid format:", quadrangle)

        cv2.polylines(frame, contours, True, (0, 255, 0), 2)
    else:
        print("No Text Detected.")

    cv2.imshow("Text Detection and Recognition", frame)
    cv2.waitKey(0)


if __name__ == "__main__":
    main()
